edsnlp/pipelines/misc/measurements/measurements.py

```diff
@@ -0,0 +1,634 @@
+        """
+        Set extensions for the measurements pipeline.
+        """
+
+        if not Span.has_extension("value"):
+            Span.set_extension("value", default=None)
+
+    def extract_units(self, term_matches: Iterable[Span]) -> Iterable[Span]:
+        """
+        Extracts unit spans from the document by extracting unit atoms (declared in the
+        units_config parameter)
+        and aggregating them automatically
+
+        Parameters
+        ----------
+        term_matches: Iterable[Span]
+
+        Returns
+        -------
+        Iterable[Span]
+        """
+        last = None
+        units = []
+        current = []
+        unit_label_hashes = set()
+        for unit_part in filter_spans(term_matches):
+            if unit_part.label not in self.unit_part_label_hashes:
+                continue
+            if last is not None and unit_part.start != last.end and len(current):
+                doc = current[0].doc
+                end = next(
+                    (i for i, e in list(enumerate(current))[::-1] if e.label_ != "per"),
+                    None,
+                )
+                if end is not None:
+                    unit = "_".join(part.label_ for part in current[: end + 1])
+                    units.append(Span(doc, current[0].start, current[end].end, unit))
+                    unit_label_hashes.add(units[-1].label)
+                current = []
+                last = None
+            if len(current) > 0 or unit_part.label_ != "per":
+                current.append(unit_part)
+            last = unit_part
+
+        end = next(
+            (i for i, e in list(enumerate(current))[::-1] if e.label_ != "per"), None
+        )
+        if end is not None:
+            doc = current[0].doc
+            unit = "_".join(part.label_ for part in current[: end + 1])
+            units.append(Span(doc, current[0].start, current[end].end, unit))
+            unit_label_hashes.add(units[-1].label)
+
+        return units
+
+    @staticmethod
+    def make_pseudo_sentence(doc, matches, pseudo_mapping):
+        """
+        Creates a pseudo sentence (one letter per entity)
+        to extract higher order patterns
+
+        Parameters
+        ----------
+        doc: Doc
+        matches: List[(Span, bool)]
+        pseudo_mapping: Dict[int, str]
+
+        Returns
+        -------
+        (str, List[int])
+        """
+        pseudo = []
+        last = 0
+        pseudo_offsets = []
+        for ent, is_sent_split in matches:
+            if ent.start != last:
+                pseudo.append("w")
+            pseudo_offsets.append(len(pseudo))
+            if is_sent_split:
+                pseudo.append(".")
+            else:
+                pseudo.append(pseudo_mapping.get(ent.label, "w"))
+            last = ent.end
+        if len(doc) != last:
+            pseudo.append("w")
+        pseudo = "".join(pseudo)
+
+        return pseudo, pseudo_offsets
+
+    def process(self, doc):
```

Comment:
This method seems overly complex. Is there a way we could simplify it, perhaps by making it more modular?

Reply:
Most of the main loop is not easily to modularizable because of the multiple local variables, but I can move out some of the matches preprocessing indeed